### PR TITLE
agent: switch to async runtime

### PR DESCRIPTION
**Pull request by mxpv**
See https://github.com/kata-containers/kata-containers/issues/1209

This commit includes minimal changes in order to switch to Tokio:
- Update protocol crate to generate async server code
- Adds async entry point to the Agent
- Updates agent services signatures in rpc.rs to support async

Signed-off-by: Maksym Pavlenko <pavlenko.maksym@gmail.com>

### CODE REVIEW

**Comment by Tim-Zhang** [comment]
@mxpv commits pushed and conflicts resolved

The current version is working now with a known problem: debug console does not working and need to async.

**Comment by Tim-Zhang** [comment]
@mxpv Do you have time to fix the debug console problem?

**Comment by mxpv** [comment]
> debug console does not working and need to async.

@Tim-Zhang For now I was able to make it work via `spawn_blocking`.
I believe it worth to rewrite debug console code because blocking `accept` calls/infinite loop/`fork()` don't work well in async ecosystem (infinite loop with single threaded pool just blocks everything and fork() may cause race conditions). But it deserves a separate PR as fork refactoring / vsock changes are not trivial and require more time.

**Comment by Tim-Zhang** [comment]
@mxpv Good job, I agree with you that we should do the async job about `debug console` in a new pr. 

After testing I found that the `debug console` works well even without the `spawn_blocking` commit. So I correct my previous comment:
image

**Comment by Tim-Zhang** [comment]
Now we reduced the thread number of agent from 17 to 4(1 for the runtime, 1 for the async logger, 2 for debug console), After we async the debug console the number will reduce to 2.

**Review by jodh-intel — COMMENTED** [review]
Thanks for raising @mxpv! This is a big PR so I'll prolly take another look later today / tomorrow but here are a few initial comments.

Async looks like a win, but it is going to mean reworking the agent tracing PR substantially (https://github.com/kata-containers/kata-containers/pull/411, but see also https://github.com/kata-containers/kata-containers/issues/60). That said, I suspect it may simplify the shutdown and tracing code even more. Would you be interested in helping out with those items?

**Comment by mxpv** [comment]
Rebased with 2.0-dev branch to fix Cargo.lock conflicts
Added headers.

File: src/agent/src/device.rs
Comment (mxpv) [reply src/agent/src/device.rs:128]: Yes, it's pretty similar (select whichever task completes first).

File: src/agent/src/device.rs
Comment (mxpv) [reply src/agent/src/device.rs:398]: This is @Tim-Zhang part, but I'll add my two cents :) 
- Because those two functions are not async (hence there is no `.await` tail).
Essentially we don't have to make every function async (unless we need to do async calls inside and make tokio to take care of thread management / state machine). In our case there is plain rust code with no async calls inside, so this can be regular funcs.
- I guess because we mix async and non async funcs, that might be hard to do.

**Comment by mxpv** [comment]
> Async looks like a win, but it is going to mean reworking the agent tracing PR substantially (#411, but see also #60). That said, I suspect it may simplify the shutdown and tracing code even more. Would you be interested in helping out with those items?

Absolutely. I'd be happy to help with this.

**Comment by jodh-intel** [comment]
Thanks for the offer @mxpv ! :smile:

File: src/agent/src/device.rs
Comment (Tim-Zhang) [reply src/agent/src/device.rs:128]: I use tokio::time::delay_for + tokio::select to instead of std receiver's `recv_timeout` because there is no similar method in tokio

**Comment by Tim-Zhang** [comment]
@jodh-intel @mxpv building failed caused by master rebasing, I will fix this tomorrow.

@mxpv Thank you for adding copyright for us but it was the old copyright, I will update it later.

File: src/agent/src/device.rs
Comment (Tim-Zhang) [reply src/agent/src/device.rs:398]: Maybe I should make these functions async for consistency reason, let me think about it.

File: src/agent/rustjail/src/filestream.rs
Comment (Tim-Zhang) [reply src/agent/rustjail/src/filestream.rs:6]: fixed

File: src/agent/rustjail/src/sync_with_async.rs
Comment (Tim-Zhang) [reply src/agent/rustjail/src/sync_with_async.rs:6]: fixed

File: src/agent/Cargo.toml
Comment (jodh-intel) [inline src/agent/Cargo.toml:14]: Can you also update `tools/agent-ctl/Cargo.toml` (currently fails to build with this PR).

**Review by jodh-intel — COMMENTED** [review]
Thanks @mxpv and @Tim-Zhang - a few more comments.

File: src/agent/src/device.rs
Comment (Tim-Zhang) [reply src/agent/src/device.rs]: I have changed these functions all to async function.

File: src/agent/src/mount.rs
Comment (Tim-Zhang) [reply src/agent/src/mount.rs]: I have changed these functions all to async function.

Comment (jodh-intel) [reply src/agent/src/mount.rs]: Thanks @Tim-Zhang

File: src/agent/src/mount.rs
Comment (Tim-Zhang) [reply src/agent/src/mount.rs:132]: it should be done in a new pr.

File: src/agent/rustjail/src/process.rs
Comment (Tim-Zhang) [reply src/agent/rustjail/src/process.rs]: I have rename the key to stream_type
Stream -> StreamType
key -> stream_type

File: src/agent/Cargo.toml
Comment (Tim-Zhang) [reply src/agent/Cargo.toml]: done

File: src/agent/rustjail/src/cgroups/notifier.rs
Comment (Tim-Zhang) [reply src/agent/rustjail/src/cgroups/notifier.rs]: done

File: src/agent/rustjail/src/cgroups/notifier.rs
Comment (Tim-Zhang) [reply src/agent/rustjail/src/cgroups/notifier.rs]: fixed

File: src/agent/src/device.rs
Comment (Tim-Zhang) [reply src/agent/src/device.rs]: done

File: src/agent/rustjail/src/cgroups/notifier.rs
Comment (Tim-Zhang) [reply src/agent/rustjail/src/cgroups/notifier.rs]: removed

File: src/agent/rustjail/src/cgroups/notifier.rs
Comment (Tim-Zhang) [reply src/agent/rustjail/src/cgroups/notifier.rs:178]: Unlike `std channel` has an infinite buffer, the `tokio channel` requires a buffer size, follow the example I use 100, Actually I don’t know what value to set, I just think 100 is enough.

File: src/agent/src/main.rs
Comment (Tim-Zhang) [reply src/agent/src/main.rs:128]: I think we can keep these functions in main.rs because it's small, or we can separate it later? when the vsock exporter use them.

File: src/agent/src/main.rs
Comment (Tim-Zhang) [reply src/agent/src/main.rs:90]: Yes it is ugly :sweat_smile: , Can you have any suggestions?

Comment (jodh-intel) [reply src/agent/src/main.rs:90]: Sorry @Tim-Zhang - I was just joking (too much sugar in my coffee! :smile:) This is fine with me.

File: src/agent/rustjail/src/sync_with_async.rs
Comment (Tim-Zhang) [reply src/agent/rustjail/src/sync_with_async.rs:4]: done

File: src/agent/src/main.rs
Comment (Tim-Zhang) [reply src/agent/src/main.rs:90]: ok, haha

File: src/agent/Cargo.toml
Comment (jodh-intel) [inline src/agent/Cargo.toml:26]: This is an old version - the latest is `"1.0.1"`, and the feature names have also changed fwics.

Comment (Tim-Zhang) [reply src/agent/Cargo.toml:26]: I plan to merge this pr first and then upgrade tokio in a new pr, because many dependent libraries need to be upgraded, which takes a lot of time. I think we can introduce the async runtime first and then we enhance it.

Comment (mxpv) [reply src/agent/Cargo.toml:26]: I was looking on what it takes to use latest version of Tokio. tokio-vsock and ttrpc needs to be upgraded to 1.0 first.

Comment (Tim-Zhang) [reply src/agent/Cargo.toml:26]: Yes, I have already created an issue for `tokio-vsock` and will update ttrpc in days.

Comment (Tim-Zhang) [reply src/agent/Cargo.toml:26]: and the netlink-sys also needs to be upgraded

Comment (jodh-intel) [reply src/agent/Cargo.toml:26]: If we can't update on this PR, please could one of you open a tracking issue so we don't forget (and ideally reference that GH issue URL in the `Cargo.toml` :smile:

Comment (Tim-Zhang) [reply src/agent/Cargo.toml:26]: @mxpv The pr for `tokio-vsock` to upgrade `tokio` https://github.com/fsyncd/tokio-vsock/pull/5

**Review by lifupan — APPROVED** [review]

File: src/agent/Cargo.toml
Comment (mxpv) [reply src/agent/Cargo.toml:26]: > many dependent libraries need to be upgraded

I've opened a tracking issue: https://github.com/kata-containers/kata-containers/issues/1257

**Comment by Tim-Zhang** [comment]
@mxpv Do you have any updates?
@jodh-intel Would you please approve this pr? if the pr is ok.

File: src/agent/Cargo.toml
Comment (Tim-Zhang) [reply src/agent/Cargo.toml:26]: @mxpv ok, thank you

**Comment by mxpv** [comment]
> @mxpv Do you have any updates?

I do, but let's have this in first, it's already big enough and hard to keep it in a green state.

File: src/agent/rustjail/src/container.rs
Comment (liubin) [inline src/agent/rustjail/src/container.rs:868]: these `close()` calls can be deleted safely?

File: src/agent/rustjail/src/container.rs
Comment (liubin) [inline src/agent/rustjail/src/container.rs:997]: can this error be keeped?

Comment (Tim-Zhang) [reply src/agent/rustjail/src/container.rs:997]: ok, will update this

File: src/agent/rustjail/src/container.rs
Comment (Tim-Zhang) [reply src/agent/rustjail/src/container.rs:868]: The PipeStream will close the fd when dropped

File: src/agent/rustjail/src/process.rs
Comment (liubin) [inline src/agent/rustjail/src/process.rs:67]: What do these `readers` and `writers` are used for?

File: src/agent/rustjail/src/sync.rs
Comment (liubin) [inline src/agent/rustjail/src/sync.rs:17]: Should this file to be delete? since you have move all the functions to `sync_with_async.rs`

**Review by liubin — APPROVED** [review]

File: src/agent/rustjail/src/sync.rs
Comment (Tim-Zhang) [reply src/agent/rustjail/src/sync.rs:17]: no, the child process is still be sync.

File: src/agent/rustjail/src/process.rs
Comment (Tim-Zhang) [reply src/agent/rustjail/src/process.rs:67]: I convert fds of stdin/stdout/... to readers and writers, store them in struct process

File: src/agent/rustjail/src/container.rs
Comment (Tim-Zhang) [reply src/agent/rustjail/src/container.rs:997]: done

**Comment by Tim-Zhang** [comment]
@mxpv @jodh-intel It's ready to merge now.

**Review by jodh-intel — APPROVED** [review]
Thanks @mxpv and @Tim-Zhang !

lgtm

**Comment by jodh-intel** [comment]
There are a lot of pending reviews on this PR - any one else want to tal before this lands...?

**Pull request merged**
